File: Convert_Sorted_Array_to_Binary_Search_Tree.py
'''
Created on 2014.8.1


Convert_Sorted_Array_to_Binary_Search_Tree

Given an array where elements are sorted in ascending order, convert it to a height balanced BST.

'''

# Definition for a  binary tree node

class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    # @param num, a list of integers
    # @return a tree node
    def sortedArrayToBST(self, num):
        if num == None or len(num) == 0:
            return None
        if len(num) == 1:
            return TreeNode(num[0]);
        else:
            length = len(num)
            root = TreeNode(num[length/2])
            root.left = self.sortedArrayToBST(num[:length/2])
            root.right = self.sortedArrayToBST(num[length/2+1:])
            return root
            
    
test = Solution()
num = [1,2,3,4,5]
root = test.sortedArrayToBST(num)
queue = [root]
while len(queue)!=0:
    node = queue[0]
    queue = queue[1:]
    print node.val
    if node.left != None:
        queue.append(node.left)
    if node.right != None:
        queue.append(node.right)

File: Cplusplus/Convert_Sorted_Array_to_Binary_Search_Tree.cpp
/*
Leetcode : Convert_Sorted_Array_to_Binary_Search_Tree.cpp
Given an array where elements are sorted in ascending order, convert it to a height balanced BST.
*/

/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode *sortedArrayToBST(vector<int> &num) {
        if(num.size() == 0)
            return NULL;
        int left = 0;
        int right = num.size() - 1;
        TreeNode *root = buildTree(num , left , right );
        return root;
    }
    TreeNode *buildTree(vector<int> &num , int left , int right )
    {
        int mid = (left + right)/2;
        TreeNode *root = new TreeNode(num[mid]);
        if(left <= mid - 1)
        {
            root->left = buildTree(num , left , mid - 1 );
        }
        if(right >= mid + 1)
        {
            root->right = buildTree(num , mid + 1 , right );
        }
        return root;
    }
};
